File: electron/steamAuthService.js
// Steam OpenID Authentication Service
// Uses system browser + local HTTP server for authentication

import { shell } from 'electron';
import http from 'http';
import { URL } from 'url';

// Steam OpenID endpoint
const STEAM_OPENID_URL = 'https://steamcommunity.com/openid/login';

// Local callback server port
const CALLBACK_PORT = 27893;
const RETURN_TO = `http://localhost:${CALLBACK_PORT}/callback`;
const REALM = `http://localhost:${CALLBACK_PORT}`;

/**
 * Extracts Steam ID64 from the OpenID claimed_id
 * @param {string} claimedId - The claimed_id from Steam OpenID response
 * @returns {string|null} - The Steam ID64 or null if invalid
 */
function extractSteamId(claimedId) {
    // claimed_id format: https://steamcommunity.com/openid/id/76561198012345678
    const match = claimedId.match(/\/openid\/id\/(\d+)$/);
    return match ? match[1] : null;
}

/**
 * Builds the Steam OpenID authentication URL
 * @returns {string} - The complete OpenID authentication URL
 */
function buildAuthUrl() {
    const params = new URLSearchParams({
        'openid.ns': 'http://specs.openid.net/auth/2.0',
        'openid.mode': 'checkid_setup',
        'openid.return_to': RETURN_TO,
        'openid.realm': REALM,
        'openid.identity': 'http://specs.openid.net/auth/2.0/identifier_select',
        'openid.claimed_id': 'http://specs.openid.net/auth/2.0/identifier_select'
    });

    return `${STEAM_OPENID_URL}?${params.toString()}`;
}

/**
 * Validates the OpenID response from Steam
 * @param {URLSearchParams} params - The callback URL parameters
 * @returns {Promise<boolean>} - Whether the response is valid
 */
async function validateOpenIdResponse(params) {
    // Create validation request
    const validationParams = new URLSearchParams();

    // Copy all openid params and change mode to check_authentication
    for (const [key, value] of params.entries()) {
        if (key.startsWith('openid.')) {
            validationParams.set(key, value);
        }
    }
    validationParams.set('openid.mode', 'check_authentication');

    try {
        const response = await fetch(STEAM_OPENID_URL, {
            method: 'POST',
            headers: {
                'Content-Type': 'application/x-www-form-urlencoded'
            },
            body: validationParams.toString()
        });

        const text = await response.text();
        return text.includes('is_valid:true');
    } catch (error) {
        console.error('OpenID validation error:', error);
        return false;
    }
}

/**
 * Creates a success HTML page
 */
function getSuccessHTML() {
    return `
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>ORBIT - Autenticación exitosa</title>
    <style>
        body {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
            background: linear-gradient(135deg, #0a0a0f 0%, #1a1a2e 100%);
            color: white;
            display: flex;
            align-items: center;
            justify-content: center;
            min-height: 100vh;
            margin: 0;
            text-align: center;
        }
        .container {
            padding: 40px;
        }
        .icon {
            font-size: 64px;
            margin-bottom: 20px;
        }
        h1 {
            color: #66c0f4;
            margin-bottom: 10px;
        }
        p {
            color: #8b949e;
        }
    </style>
</head>
<body>
    <div class="container">
        <div class="icon">✓</div>
        <h1>¡Autenticación exitosa!</h1>
        <p>Tu cuenta de Steam ha sido vinculada a ORBIT.</p>
        <p>Puedes cerrar esta ventana.</p>
    </div>
</body>
</html>`;
}

/**
 * Creates an error HTML page
 */
function getErrorHTML(error) {
    return `
<!DOCTYPE html>
<html>
<head>
    <meta charset="UTF-8">
    <title>ORBIT - Error de autenticación</title>
    <style>
        body {
            font-family: -apple-system, BlinkMacSystemFont, 'Segoe UI', Roboto, sans-serif;
            background: linear-gradient(135deg, #0a0a0f 0%, #1a1a2e 100%);
            color: white;
            display: flex;
            align-items: center;
            justify-content: center;
            min-height: 100vh;
            margin: 0;
            text-align: center;
        }
        .container {
            padding: 40px;
        }
        .icon {
            font-size: 64px;
            margin-bottom: 20px;
        }
        h1 {
            color: #ef4444;
            margin-bottom: 10px;
        }
        p {
            color: #8b949e;
        }
    </style>
</head>
<body>
    <div class="container">
        <div class="icon">✕</div>
        <h1>Error de autenticación</h1>
        <p>${error}</p>
        <p>Puedes cerrar esta ventana e intentarlo de nuevo.</p>
    </div>
</body>
</html>`;
}

/**
 * Starts the Steam OpenID authentication flow
 * Opens system browser for Steam login and waits for callback
 * @returns {Promise<{success: boolean, steamId?: string, error?: string}>}
 */
async function startSteamAuth() {
    return new Promise((resolve) => {
        let resolved = false;
        let server = null;

        // Timeout after 5 minutes
        const timeout = setTimeout(() => {
            if (!resolved) {
                resolved = true;
                if (server) server.close();
                resolve({ success: false, error: 'Tiempo de espera agotado' });
            }
        }, 5 * 60 * 1000);

        // Create local HTTP server to capture callback
        server = http.createServer(async (req, res) => {
            const url = new URL(req.url, `http://localhost:${CALLBACK_PORT}`);

            if (url.pathname === '/callback') {
                const params = url.searchParams;
                const mode = params.get('openid.mode');

                if (mode === 'cancel') {
                    res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
                    res.end(getErrorHTML('Autenticación cancelada'));

                    if (!resolved) {
                        resolved = true;
                        clearTimeout(timeout);
                        server.close();
                        resolve({ success: false, error: 'Autenticación cancelada' });
                    }
                    return;
                }

                if (mode === 'id_res') {
                    // Validate the response
                    const isValid = await validateOpenIdResponse(params);

                    if (!isValid) {
                        res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
                        res.end(getErrorHTML('Respuesta de Steam no válida'));

                        if (!resolved) {
                            resolved = true;
                            clearTimeout(timeout);
                            server.close();
                            resolve({ success: false, error: 'Respuesta de Steam no válida' });
                        }
                        return;
                    }

                    // Extract Steam ID
                    const claimedId = params.get('openid.claimed_id');
                    const steamId = extractSteamId(claimedId);

                    if (!steamId) {
                        res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
                        res.end(getErrorHTML('No se pudo obtener el Steam ID'));

                        if (!resolved) {
                            resolved = true;
                            clearTimeout(timeout);
                            server.close();
                            resolve({ success: false, error: 'No se pudo obtener el Steam ID' });
                        }
                        return;
                    }

                    // Success!
                    res.writeHead(200, { 'Content-Type': 'text/html; charset=utf-8' });
                    res.end(getSuccessHTML());

                    if (!resolved) {
                        resolved = true;
                        clearTimeout(timeout);
                        server.close();
                        resolve({ success: true, steamId });
                    }
                    return;
                }
            }

            // Default response
            res.writeHead(404);
            res.end('Not Found');
        });

        server.on('error', (err) => {
            console.error('Server error:', err);
            if (!resolved) {
                resolved = true;
                clearTimeout(timeout);
                resolve({ success: false, error: `Error del servidor: ${err.message}` });
            }
        });

        // Start server and open browser
        server.listen(CALLBACK_PORT, () => {
            console.log(`Steam auth callback server listening on port ${CALLBACK_PORT}`);

            // Open Steam login in system browser
            const authUrl = buildAuthUrl();
            console.log('Opening Steam auth URL in system browser:', authUrl);
            shell.openExternal(authUrl);
        });
    });
}

export const steamAuthService = {
    startSteamAuth,
    extractSteamId,
    validateOpenIdResponse
};

File: Orbit.Core/Services/SteamAuthService.cs
using System;
using System.Diagnostics;
using System.Net;
using System.Threading.Tasks;
using System.Web;
using System.Collections.Specialized;
using System.Text.Json;

namespace Orbit.Core.Services
{
    public class SteamAuthService
    {
        private const string OPENID_URL = "https://steamcommunity.com/openid/login";
        private const string CALLBACK_URL = "http://127.0.0.1:5000/auth/steam/return";

        public async Task<string?> StartSteamAuth()
        {
            // 1. Construct OpenID Paramaters
            var query = HttpUtility.ParseQueryString(string.Empty);
            query["openid.ns"] = "http://specs.openid.net/auth/2.0";
            query["openid.mode"] = "checkid_setup";
            query["openid.return_to"] = CALLBACK_URL;
            query["openid.realm"] = "http://127.0.0.1:5000";
            query["openid.identity"] = "http://specs.openid.net/auth/2.0/identifier_select";
            query["openid.claimed_id"] = "http://specs.openid.net/auth/2.0/identifier_select";

            var authUrl = $"{OPENID_URL}?{query}";

            // 2. Start Local HTTP Server for Callback
            var listener = new HttpListener();
            listener.Prefixes.Add("http://127.0.0.1:5000/");
            listener.Start();

            // 3. Open System Browser
            Process.Start(new ProcessStartInfo
            {
                FileName = authUrl,
                UseShellExecute = true
            });

            try
            {
                // 4. Wait for Callback
                var context = await listener.GetContextAsync();
                var request = context.Request;
                var response = context.Response;

                // 5. Validate parameters 
                // (In a real app, strict validation against Steam server is required 'check_authentication')
                // For now, parsing the claimed_id is enough for MVP

                var steamId = ExtractSteamId(request.QueryString["openid.claimed_id"]);

                // 6. Send Response to Browser
                string responseString = @"
<!DOCTYPE html>
<html lang='es'>
<head>
    <meta charset='UTF-8'>
    <meta name='viewport' content='width=device-width, initial-scale=1.0'>
    <title>Orbit - Conexión Exitosa</title>
    <style>
        body {
            background-color: #0f172a; /* Slate 900 */
            color: #e2e8f0; /* Slate 200 */
            font-family: 'Segoe UI', system-ui, sans-serif;
            display: flex;
            align-items: center;
            justify-content: center;
            height: 100vh;
            margin: 0;
        }
        .card {
            background: #1e293b; /* Slate 800 */
            padding: 2rem 3rem;
            border-radius: 1rem;
            box-shadow: 0 10px 15px -3px rgba(0, 0, 0, 0.5);
            text-align: center;
            max-width: 400px;
            width: 100%;
            border: 1px solid #334155;
        }
        h1 {
            color: #fff;
            margin-bottom: 0.5rem;
            font-size: 1.5rem;
        }
        p {
            color: #94a3b8;
            margin-bottom: 2rem;
        }
        .icon {
            width: 80px;
            height: 80px;
            background: linear-gradient(135deg, #6366f1 0%, #8b5cf6 100%);
            border-radius: 50%;
            display: flex;
            align-items: center;
            justify-content: center;
            margin: 0 auto 1.5rem;
            box-shadow: 0 0 20px rgba(99, 102, 241, 0.3);
        }
        .icon svg {
            width: 40px;
            height: 40px;
            color: white;
            stroke-width: 3;
        }
        .btn {
            background: #334155;
            color: white;
            border: none;
            padding: 0.75rem 1.5rem;
            border-radius: 0.5rem;
            cursor: pointer;
            font-size: 0.9rem;
            transition: all 0.2s;
        }
        .btn:hover {
            background: #475569;
        }
    </style>
</head>
<body>
    <div class='card'>
        <div class='icon'>
            <svg xmlns='http://www.w3.org/2000/svg' fill='none' viewBox='0 0 24 24' stroke='currentColor'>
                <path stroke-linecap='round' stroke-linejoin='round' d='M4.5 12.75l6 6 9-13.5' />
            </svg>
        </div>
        <h1>¡Conectado con Steam!</h1>
        <p>Tu cuenta se ha vinculado correctamente. Ya puedes cerrar esta ventana y volver a Orbit.</p>
        <button class='btn' onclick='window.close()'>Cerrar ventana</button>
    </div>
    <script>
        // Try to auto-close after 3 seconds
        setTimeout(() => {
            window.close();
        }, 3000);
    </script>
</body>
</html>";
                byte[] buffer = System.Text.Encoding.UTF8.GetBytes(responseString);
                response.ContentLength64 = buffer.Length;
                response.OutputStream.Write(buffer, 0, buffer.Length);
                response.OutputStream.Close();

                return steamId;
            }
            catch (Exception ex)
            {
                Console.WriteLine(ex.Message);
                return null;
            }
            finally
            {
                listener.Stop();
            }
        }

        private string? ExtractSteamId(string? claimedId)
        {
            if (string.IsNullOrEmpty(claimedId)) return null;
            // Format: https://steamcommunity.com/openid/id/76561198...
            var parts = claimedId.Split('/');
            return parts[parts.Length - 1];
        }
    }
}
